crates/metadata-store/src/rocksdb/server.rs

```diff
@@ -0,0 +1,344 @@
+                }
+                _ = cancellation_watcher() => {
+                    break;
+                },
+            }
+        }
+
+        debug!("Stopped RocksDBMetadataStore");
+    }
+
+    fn kv_cf_handle(&self) -> &ColumnFamily {
+        self.db
+            .cf_handle(KV_PAIRS)
+            .expect("KV_PAIRS column family exists")
+    }
+
+    async fn handle_request(&self, caller: NodeId, request: MetadataStoreRequest) {
+        trace!("Handle request '{:?}' from node '{}'", request, caller);
+
+        let MetadataStoreRequest { request_id, kind } = &request;
+
+        let response = match kind {
+            RequestKind::Get(Get { key }) => ResponseKind::GetResponse(
+                Self::log_error(self.get(key), &request).map_err(Into::into),
+            ),
+            RequestKind::GetVersion(GetVersion { key }) => ResponseKind::GetVersionResponse(
+                Self::log_error(self.get_version(key), &request).map_err(Into::into),
+            ),
+            RequestKind::Put(Put {
+                key,
+                value,
+                precondition,
+            }) => ResponseKind::PutResponse(
+                Self::log_error(self.put(key, value, precondition), &request).map_err(Into::into),
+            ),
+            RequestKind::Delete(Delete { key, precondition }) => ResponseKind::DeleteResponse(
+                Self::log_error(self.delete(key, precondition), &request).map_err(Into::into),
+            ),
+        };
+
+        let _ = task_center().spawn_child(
+            TaskKind::Disposable,
+            "send-metadata-store-response-to-client",
+            None,
+            {
+                let networking = self.networking.clone();
+                let request_id = *request_id;
+                async move {
+                    networking
+                        .send(
+                            caller,
+                            &MetadataStoreResponse {
+                                request_id,
+                                kind: response,
+                            },
+                        )
+                        .await?;
+                    Ok(())
+                }
+            },
+        );
+    }
+
+    fn get(&self, key: &ByteString) -> Result<Option<VersionedValue>> {
+        let cf_handle = self.kv_cf_handle();
+        let slice = self.db.get_pinned_cf(cf_handle, key)?;
+
+        if let Some(bytes) = slice {
+            Ok(Some(Self::decode(bytes)?))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn get_version(&self, key: &ByteString) -> Result<Option<Version>> {
+        let cf_handle = self.kv_cf_handle();
+        let slice = self.db.get_pinned_cf(cf_handle, key)?;
+
+        if let Some(bytes) = slice {
+            let versioned_value = Self::decode::<VersionedValue>(bytes)?;
+            Ok(Some(versioned_value.version))
+        } else {
+            Ok(None)
+        }
+    }
```

Comment:
It would have been 👏 if we only deserialized the version portion of the payload instead of deserializing the entire value, but admittedly, it's an optimization.

Reply:
Maybe we can tackle this aspect when introducing a properly versioned format.

Reply:
I'll add a todo to record it.